### ISSUE TITLE
Frontmatter extension for Qute

### ISSUE THREAD

**Issue by ia3andy** [quarkusio/quarkus#33251]
### Description

Interested in this extension, please +1 via the emoji/reaction feature of GitHub (top right).

```html
{#include main.html}
{#title}My Notes{/title}
<p>Hello</p>
{/include}
```

could be replaced by:
```html
include: main.html
title: My Notes
---
<p>Hello</p>
```

### Repository name

quarkus-qute-frontmatter

### Short description

Allow to use frontmatter alongside Qute to make layout a bit simpler

### Repository Homepage URL

https://quarkiverse.github.io/quarkiverse-docs/quarkus-qute-frontmatter/dev/

### Repository Topics

- quarkus-extension
- qute
- template
- frontmatter
- layout

### Team Members

- @ia3andy 
- @mkouba 

### Additional context

Zulip discussion: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Frontmatter.20extension.20for.20Qute

**Comment by ia3andy** [comment]
> Why is frontmatter something to tie into qute? isn't it something relevant to apply to all kinds of content ? HTML, markdown, liquid, handlebar etc.?

Those would be different extension as to be able to do this you have to tie it to the underlying template engine AFAIK

**Comment by maxandersen** [comment]
> Those would be different extensions as to be able to do this you have to tie it to the underlying template engine AFAIK

yes - hence its not really a qute thing; but a "blog-generator" thing.

Like, what is the usecase for the notion of frontmatters *only* on qute templates? and would it fit into quarkus typesafe templates etc?

**Comment by ia3andy** [comment]
> > Those would be different extensions as to be able to do this you have to tie it to the underlying template engine AFAIK
> 
> yes - hence its not really a qute thing; but a "blog-generator" thing.
> 
> Like, what is the usecase for the notion of frontmatters _only_ on qute templates? and would it fit into quarkus typesafe templates etc?

for the first question I think other frontmatter extension could also make sense, but Qute is the one I use with Quarkus.

For the secodn question about type safety, @mkouba might know more

**Comment by mkouba** [comment]
>>  and would it fit into quarkus typesafe templates etc?

> For the secodn question about type safety, @mkouba might know more

As I explained in the zulip discussion - I think that the best way to implement this would be to add a custom [content filter](https://quarkus.io/guides/qute-reference#content-filters) that would modify the template before it's parsed. Basically, replace the frontmatter with qute tags. This should not affect type-safe templates. However, keep in mind that similar solutions can be brittle...

**Comment by FroMage** [comment]
I already proposed a shorter syntax in https://github.com/quarkusio/quarkus/issues/8750 a long time ago, and lost that battle:

```
{#set title}My Title{/} 
{#extend base/}
<div>
  My body.
</div>
```

That looks more Qute-spirited and shorter than frontmatter ;)

**Comment by maxandersen** [comment]
@FroMage +1 - expanding/simplifying the syntax of Qute's ability to composite templates makes more sense to me than adding a "flat" frontmatter.

**Comment by mkouba** [comment]
> I already proposed a shorter syntax in #8750 a long time ago, and lost that battle:
> 
> ```
> {#set title}My Title{/} 
> {#extend base/}
> <div>
>   My body.
> </div>
> ```
> 
> That looks more Qute-spirited and shorter than frontmatter ;)

Well, I don't find this proposal more readable or even Qute-spirited and it definitely does not follow the current syntax rules. 

And BTW this particular example could be simplified like:
```
{#include base title='My Title'}
<div>...  // this would be used for {#insert}
{/}
```
and then in the base `<title>{title ?: 'Default title'}</title> `.

**Comment by ia3andy** [comment]
> 

Ah that could be a solution indeed (for both problems), I would auto-add the end tag at the end of the current section though (not the end of the template)

**Comment by ia3andy** [comment]
I would be happy with such a happy ending :)

**Comment by mkouba** [comment]
> I would auto-add the end tag at the end of the current section though (not the end of the template)

:+1:

**Comment by mkouba** [comment]
FYI I've created https://github.com/quarkusio/quarkus/issues/33293
